### PR TITLE
Future Stencil + Shared GTCaches

### PR DESCRIPTION
**Pull request by eddie-c-davis**
## Purpose

This PR introduces the `future_stencil` decorator function and corresponding `FutureStencil` class. The purpose of these two constructs is to enable distributed compilation. This is accomplished via a distributed stencil table that uses one sided MPI communication to coordinate which node (rank) is responsible for compiling each stencil. Other nodes wait (blocking) for the needed stencil to be compiled before loading and continuing with execution. This allows all compiled stencils to be stored in a single shared stencil cache on disk.

## Code changes:

- The code in `gt4py_utils` that creates separate `.gt_cache` directories for each rank is removed
- `FrozenStencil` is modified to call the `gtscript.stencil` directory to compile stencils in a sequential context or the `fv3core.future_stencil` decorator in a parallel context.
- The `future_stencil`decorator function and `FutureStencil` class are added
- Unit tests are added for one sided MPI communication, the distributed stencil table, deferred compilation, and distributed compilation in `test_future_stencil`

## Infrastructure changes:

- This will require changes to the saved stencil caches for the Jenkins plans as code will no longer be stored in per node stencil caches

## Checklist
Before submitting this PR, please make sure:

- [x] You have followed the coding standards guidelines established at [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--BD7zigBMAhMZAPkeNENeuU2UAg-IlsYffZgTwyKEylty7NhY).
- [x] Docstrings and type hints are added to new and updated routines, as appropriate
- [x] Unit tests are added or updated for non-stencil code changes

### CODE REVIEW

**Comment by eddie-c-davis** [comment]
launch jenkins

**Comment by eddie-c-davis** [comment]
launch jenkins

**Comment by eddie-c-davis** [comment]
launch jenkins

**Comment by eddie-c-davis** [comment]
launch jenkins

**Comment by eddie-c-davis** [comment]
launch jenkins

**Comment by eddie-c-davis** [comment]
launch jenkins

**Comment by eddie-c-davis** [comment]
launch jenkins

**Comment by eddie-c-davis** [comment]
launch jenkins

File: fv3core/utils/future_stencil.py [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+        if self._stencil_object is None:
+            self._wait_for_stencil()
+        return self._stencil_object
+
+    @property
+    def field_info(self) -> Dict[str, FieldInfo]:
+        return self.stencil_object.field_info
+
+    def _delay(self) -> float:
+        delay_time = self._sleep_time * float(self._node_id)
```
Comment (FlorianDeconinck) [inline fv3core/utils/future_stencil.py:243]:
```suggestion
        delay_time = self._sleep_time + float(self._node_id)
```
With the current `_sleep_time`, node 1000 would wait 50s

File: fv3core/utils/future_stencil.py [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+
+    def __init__(
+        self,
+        builder: Optional["StencilBuilder"] = None,
+        wrapper: Optional[Callable] = None,
+    ):
+        self._builder: Optional["StencilBuilder"] = builder
+        self._stencil_object: Optional[StencilObject] = None
+        self._sleep_time: float = 0.05
+        self._timeout: float = 180.0
```
Comment (FlorianDeconinck) [inline fv3core/utils/future_stencil.py:220]: We have stencils that take more than 180s, e.g. physics. I have no good idea apart from raising that limit though :/

Comment (FlorianDeconinck) [reply fv3core/utils/future_stencil.py:220]: Technically we would need a ping to the node compiling to make sure it's alive

**Review by FlorianDeconinck — APPROVED** [review]
Some docstrings and looking at that sleep_time potential scaling issue.

Rest looks good!

File: fv3core/utils/future_stencil.py (hunk 1 above)
Comment (FlorianDeconinck) [reply fv3core/utils/future_stencil.py:243]: Actually that won't work either! Should be doing `_sleep_time*float(self._node_id) / 10` or something

File: .jenkins/jenkins.sh [hunk 3]
```diff
@@ -109,9 +109,8 @@ if grep -q "parallel" <<< "${script}"; then
 	fi
 	if [ -f ${scheduler_script} ] ; then
 	    sed -i 's|<NTASKS>|<NTASKS>\n#SBATCH \-\-hint=multithread\n#SBATCH --ntasks-per-core=2|g' ${scheduler_script}
-	    sed -i 's|45|30|g' ${scheduler_script}
 	    if [ "$NUM_RANKS" -gt "6" ];then
-            sed -i 's|cscsci|debug|g' ${scheduler_script}
```
[the remark this answers is not in the document]
Comment (eddie-c-davis) [reply .jenkins/jenkins.sh:114]: I think so, perhaps not after the delay change. Will test without it...

File: fv3core/utils/future_stencil.py (hunk 2 above)
Comment (eddie-c-davis) [reply fv3core/utils/future_stencil.py:220]: That could potentially be done, but for now I would be in favor of increasing the timeout.

File: fv3core/utils/future_stencil.py (hunk 1 above)
Comment (eddie-c-davis) [reply fv3core/utils/future_stencil.py:243]: Great idea.

File: fv3core/utils/future_stencil.py (hunk 2 above)
Comment (FlorianDeconinck) [reply fv3core/utils/future_stencil.py:220]: Can we comment that in though? So at least it's commented in-code

Comment (eddie-c-davis) [reply fv3core/utils/future_stencil.py:220]: Absolutely.

**Pull request merged**